tests/test_utils_url.py

```diff
@@ -73,6 +73,48 @@ class MySpider(Spider):
+        self.assertEqual(add_http_if_no_scheme('www.example.com'),
+                                               'http://www.example.com')
+        self.assertEqual(add_http_if_no_scheme('example.com'),
+                                               'http://example.com')
+        self.assertEqual(add_http_if_no_scheme('//example.com'),
+                                               'http://example.com')
+        self.assertEqual(add_http_if_no_scheme('//www.example.com/some/page.html'),
+                                               'http://www.example.com/some/page.html')
+        self.assertEqual(add_http_if_no_scheme('www.example.com:80'),
+                                               'http://www.example.com:80')
+        self.assertEqual(add_http_if_no_scheme('www.example.com:80/some/page.html'),
+                                               'http://www.example.com:80/some/page.html')
+        self.assertEqual(add_http_if_no_scheme('http://www.example.com:80/some/page.html'),
+                                               'http://www.example.com:80/some/page.html')
+        self.assertEqual(add_http_if_no_scheme('www.example.com/some/page#frag'),
+                                               'http://www.example.com/some/page#frag')
+        self.assertEqual(add_http_if_no_scheme('http://www.example.com/some/page#frag'),
+                                               'http://www.example.com/some/page#frag')
+        self.assertEqual(add_http_if_no_scheme('www.example.com/do?a=1&b=2&c=3'),
+                                               'http://www.example.com/do?a=1&b=2&c=3')
+        self.assertEqual(add_http_if_no_scheme('http://www.example.com/do?a=1&b=2&c=3'),
+                                               'http://www.example.com/do?a=1&b=2&c=3')
+        self.assertEqual(add_http_if_no_scheme('username:password@example.com/some/page.html'),
+                                               'http://username:password@example.com/some/page.html')
+        self.assertEqual(add_http_if_no_scheme('http://username:password@example.com/some/page.html'),
+                                               'http://username:password@example.com/some/page.html')
+        self.assertEqual(add_http_if_no_scheme('username:password@example.com:80/some/part?a=1&b=2&c=3#frag'),
+                                               'http://username:password@example.com:80/some/part?a=1&b=2&c=3#frag')
+        self.assertEqual(add_http_if_no_scheme('http://username:password@example.com:80/some/part?a=1&b=2&c=3#frag'),
+                                               'http://username:password@example.com:80/some/part?a=1&b=2&c=3#frag')
+        self.assertEqual(add_http_if_no_scheme('https://www.example.com'),
+                                               'https://www.example.com')
+        self.assertEqual(add_http_if_no_scheme('ftp://www.example.com'),
+                                               'ftp://www.example.com')
```

Comment:
Could you please split this test into several test cases? It makes sense to create a new TestCase for add_http_if_no_scheme and create several smaller test methods (e.g. `test_username_password`, `test_preserve_ftp`, `test_leading_slashes`, etc). 

I know other tests in this module don't follow this convention (yet), but we're slowly moving to more fine-grained test methods because they are easier to read and provide a possiblity to run the rest of the tests if one of them fails.